### PR TITLE
adding proxy support for aws v2

### PR DESCRIPTION
**Pull request by ajaijoy-bmc**
Created AWSProxy as springboot @configurationProperties
Updated AmazonClientProvider
Tested with and without proxy

### CODE REVIEW

**Comment by ajaijoy-bmc** [comment]
@cfieber  and @ajordens . Please let me know your thoughts.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java [hunk 1]
```diff
@@ -44,6 +52,7 @@
     private final String defaultKeyPair;
     private final List<String> requiredGroupMembership;
     private final List<AWSRegion> regions;
+    private AWSProxy proxy;
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: This should not be necessary

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java [hunk 2]
```diff
@@ -125,6 +134,178 @@ public String getDefaultKeyPair() {
         return regions;
     }
 
+    public AWSProxy getAWSProxy() {
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: should not be necessary

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java [hunk 3]
```diff
@@ -125,6 +134,178 @@ public String getDefaultKeyPair() {
...
+        return proxy;
+    }
+
+
+    @ConfigurationProperties(prefix = "aws.proxy", locations = "clouddriver.yml")
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: should be able to take the `locations` attribute off this annotation

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java [hunk 4]
```diff
@@ -125,6 +134,178 @@ public String getDefaultKeyPair() {
...
+    @Component
+    public static class AWSProxy {
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: Can you pull this class out of here to its own top level class?

Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: will do...

Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: You should be able to just make this a Groovy @Canonical class:
````groovy
class AWSProxy {
  String proxyHost
  String proxyPort
  String proxyUsername
  String proxyPassword
  String proxyDomain
  String proxyWorkstation
  String protocol
}
````
and add your extra business logic methods (`apply`, `isNTLMProxy`, `isProxyMode`)

File: clouddriver-web/config/clouddriver.yml [hunk 5]
```diff
@@ -42,6 +42,14 @@ default:
 
 aws:
   enabled: ${AWS_ENABLED:false}
+  #proxy:
```
Comment (cfieber) [inline clouddriver-web/config/clouddriver.yml]: thanks for the config sample here

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java (hunk 4 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: can i maintain it in java itself ?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 6]
```diff
@@ -58,248 +59,263 @@
...
+
+
+    protected <T extends AmazonWebServiceClient> T getClient(Class<T> impl, NetflixAmazonCredentials amazonCredentials, String region) throws IllegalAccessException, InvocationTargetException,
+            InstantiationException, NoSuchMethodException {
+        Constructor<T> constructor = impl.getConstructor(AWSCredentialsProvider.class, ClientConfiguration.class);
+
+        ClientConfiguration clientConfiguration = new ClientConfiguration();
+        clientConfiguration.setRetryPolicy(retryPolicy);
+
+        if (proxy != null && proxy.isProxyConfigMode())
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: can you put braces around this statement please ` { } `

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java (hunk 4 above)
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: sure, I don't mind.  The jackson annotations shouldn't be required for spring configuration properties binding

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java (hunk 6 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: sure will add brace for if

**Comment by cfieber** [comment]
@ajaijoy-bmc One overall comment on this change: Can you revert any lines that are unaffected by the specific change you are trying to make? It looks like you autoformated a few of the classes in here and it adds significantly more noise to the changeset

**Comment by cfieber** [comment]
Looks pretty close, thanks!

**Comment by ajaijoy** [comment]
@cfieber  Implemented feedback changes...

**Comment by ajaijoy-bmc** [comment]
@ajordens  please let me know your thoughts...

**Comment by ajaijoy** [comment]
@cfieber thoughts?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 7]
```diff
@@ -58,248 +59,264 @@
  */
 public class AmazonClientProvider {
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: The formatting changes in this file still need to be backed out

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java [hunk 8]
```diff
@@ -16,24 +16,24 @@
...
 
 /**
  * Basic set of Amazon credentials that will a provided {@link com.amazonaws.auth.AWSCredentialsProvider} to resolve account credentials.
  * If none provided, the {@link com.amazonaws.auth.DefaultAWSCredentialsProviderChain} will be used. The account's active
  * regions and availability zones can be specified as well.
- *
- *
  */
+
 public class AmazonCredentials implements AccountCredentials<AWSCredentials> {
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonCredentials.java]: looks like there are no changes in this file, please revert it back to HEAD

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy [hunk 9]
```diff
@@ -30,7 +30,9 @@ import com.netflix.spinnaker.clouddriver.aws.deploy.userdata.LocalFileUserDataPr
 import com.netflix.spinnaker.clouddriver.aws.deploy.userdata.UserDataProvider
 import com.netflix.spinnaker.clouddriver.aws.model.SecurityGroupLookupFactory
 import com.netflix.spinnaker.clouddriver.aws.provider.AwsCleanupProvider
+import com.netflix.spinnaker.clouddriver.aws.security.AWSProxy
 import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonCredentials
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy]: Does this import need to be here?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy [hunk 10]
```diff
@@ -44,12 +46,7 @@ import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
 import org.springframework.boot.context.properties.ConfigurationProperties
 import org.springframework.boot.context.properties.EnableConfigurationProperties
-import org.springframework.context.annotation.Bean
-import org.springframework.context.annotation.ComponentScan
-import org.springframework.context.annotation.Configuration
-import org.springframework.context.annotation.DependsOn
-import org.springframework.context.annotation.Import
-import org.springframework.context.annotation.Scope
+import org.springframework.context.annotation.*
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy]: can you revert this change

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java (hunk 7 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: will do.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy (hunk 9 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy]: Yes it is used in @ L80 when creating AmazonClientProvider

Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/AwsConfiguration.groovy]: `AmazonClientProvider` not `AWSProxy`

**Comment by cfieber** [comment]
LGTM - if you can rebase it one more time I will merge

**Comment by ajaijoy** [comment]
great will do now in 2 mins...

**Comment by ajaijoy** [comment]
rebased ...
